Rename ns alias improvements

yuhan0 opened the pull request:
Some minor improvements to the command `clojure-rename-ns-alias`, offering minibuffer completion of existing aliases and using regexp-quote, so that namespaces containing characters like `*` or `+` can be renamed correctly without need for manual escaping.

-----------------

Before submitting a PR mark the checkboxes for the items you've done (if you
think a checkbox does not apply, then leave it unchecked):

- [x] The commits are consistent with our [contribution guidelines][1].
- [x] You've added tests (if possible) to cover your change(s). Bugfix, indentation, and font-lock tests are extremely important!
- [x] You've run `M-x checkdoc` and fixed any warnings in the code you've written.
- [x] You've updated the changelog (if adding/changing user-visible functionality).
- [ ] You've updated the readme (if adding/changing user-visible functionality).

bbatsov commented on clojure-mode.el:
```diff
@@ -2684,6 +2684,19 @@ lists up."
     (insert sexp)
     (clojure--replace-sexps-with-bindings-and-indent)))
 
+(defun clojure--collect-ns-aliases ()
```
I guess this one can even be public API and optionally take a `ns` form as a param.

yuhan0 replied:
Do you mean having it as a string->list<string> pure function? Or taking a `point` param marking the location of the start of the ns form?

bbatsov replied:
Pure function taking a ns form as param.

bbatsov commented on clojure-mode.el:
```diff
@@ -2684,6 +2684,19 @@ lists up."
+      (while (re-search-forward rgx end 'noerror)
+        (message (match-string 1))
+        (push (cons (point) (match-string-no-properties 1)) res))
+      res)))
```
I'm not sure I understand the value of returning cons cells here if the first item in the cons cell is not the name the alias is referring to.

yuhan0 replied:
Hmm, I wrote this a few months ago and can't remember why I had that in the first place, will simplify

bbatsov replied:
👍

bbatsov commented on CHANGELOG.md:
```diff
@@ -2,8 +2,12 @@
 
-### Bugs fixed
+### New features
+* [#556](https://github.com/clojure-emacs/clojure-mode/issues/556): `clojure-rename-ns-alias` picks up existing aliases for minibuffer completion.
```
Please, add a blank line after each heading.

yuhan0 commented:
I discovered and fixed another bug when aliases have common prefixes, eg.
```clj
[clojure.string :as string]
[clojure.spec.alpha :as s]
```

renaming s -> x would produce `[clojure.string :as xtring]`

Refactored `clojure-collect-ns-aliases` into a pure fn as requested, although it does seem a little un-idiomatic for elisp.

I can squash related commits together if needed, thanks

bbatsov commented:
> Refactored clojure-collect-ns-aliases into a pure fn as requested, although it does seem a little un-idiomatic for elisp.

Looks fine to me. Add a couple of unit tests for it and we're in business. 

> I discovered and fixed another bug when aliases have common prefixes, eg.

Great!

> I can squash related commits together if needed, thanks

That'd be ideal! Thanks!

yuhan0 commented:
Okay, squashed the commits and included one test for the collect-aliases function.

bbatsov commented:
Thanks!
